.github/scripts/gitutils.py

```diff
@@ -0,0 +1,444 @@
+
+    def rev_parse(self, name: str) -> str:
+        return self._run_git("rev-parse", "--verify", name).strip()
+
+    def get_merge_base(self, from_ref: str, to_ref: str) -> str:
+        return self._run_git("merge-base", from_ref, to_ref).strip()
+
+    def patch_id(self, ref: Union[str, List[str]]) -> List[Tuple[str, str]]:
+        is_list = isinstance(ref, list)
+        if is_list:
+            if len(ref) == 0:
+                return []
+            ref = " ".join(ref)
+        rc = _check_output(
+            ["sh", "-c", f"git -C {self.repo_dir} show {ref}|git patch-id --stable"]
+        ).strip()
+        return [cast(Tuple[str, str], x.split(" ", 1)) for x in rc.split("\n")]
+
+    def commits_resolving_gh_pr(self, pr_num: int) -> List[str]:
+        owner, name = self.gh_owner_and_name()
+        msg = f"Pull Request resolved: https://github.com/{owner}/{name}/pull/{pr_num}"
+        rc = self._run_git("log", "--format=%H", "--grep", msg).strip()
+        return rc.split("\n") if len(rc) > 0 else []
+
+    def get_commit(self, ref: str) -> GitCommit:
+        return parse_fuller_format(
+            self._run_git("show", "--format=fuller", "--date=unix", "--shortstat", ref)
+        )
+
+    def cherry_pick(self, ref: str) -> None:
+        self._run_git("cherry-pick", "-x", ref)
+
+    def revert(self, ref: str) -> None:
+        self._run_git("revert", "--no-edit", ref)
+
+    def compute_branch_diffs(
+        self, from_branch: str, to_branch: str
+    ) -> Tuple[List[str], List[str]]:
+        """
+        Returns list of commmits that are missing in each other branch since their merge base
+        Might be slow if merge base is between two branches is pretty far off
+        """
+        from_ref = self.rev_parse(from_branch)
+        to_ref = self.rev_parse(to_branch)
+        merge_base = self.get_merge_base(from_ref, to_ref)
+        from_commits = self.revlist(f"{merge_base}..{from_ref}")
+        to_commits = self.revlist(f"{merge_base}..{to_ref}")
+        from_ids = fuzzy_list_to_dict(self.patch_id(from_commits))
+        to_ids = fuzzy_list_to_dict(self.patch_id(to_commits))
+        for patch_id in set(from_ids).intersection(set(to_ids)):
+            from_values = from_ids[patch_id]
+            to_values = to_ids[patch_id]
+            if len(from_values) != len(to_values):
+                # Eliminate duplicate commits+reverts from the list
+                while len(from_values) > 0 and len(to_values) > 0:
+                    frc = self.get_commit(from_values.pop())
+                    toc = self.get_commit(to_values.pop())
+                    # FRC branch might have PR number added to the title
+                    if (  # noqa: SIM102
+                        frc.title != toc.title or frc.author_date != toc.author_date
+                    ):
+                        # HACK: Same commit were merged, reverted and landed again
+                        # which creates a tracking problem
```

Comment:
Curious to ask: What is the tracking problem? Would you mind elaborating more? I think viablestrict should never set to a commit that will be reverted (because it's not green), so why is tracking the reverted/re-landed hash matters

Reply:
Let me check this part from PyTorch for more historical context, the script is copied from there.  One one hand, it's easier to keep the script as it is (which I think I will do).  On the other hand, this definitely looks like some hacks to be removed.